packages/kbn-test/src/auth/saml_auth.ts

```diff
@@ -279,20 +274,49 @@ export const finishSAMLHandshake = async ({
   };
   let authResponse: AxiosResponse;
 
-  try {
-    authResponse = await axios.request(request);
-  } catch (ex) {
-    log.error('Failed to call SAML callback');
-    cleanException(url, ex);
-    // Logging the `Cookie: sid=xxxx` header is safe here since it’s an intermediate, non-authenticated cookie that cannot be reused if leaked.
-    log.error(`Request sent: ${util.inspect(request)}`);
-    throw ex;
+  let attemptsLeft = retriesCount;
+  while (attemptsLeft > 0) {
+    try {
+      authResponse = await axios.request(request);
+
+      // SAML callback should return 302
+      if (authResponse.status === 302) {
+        return getCookieFromResponseHeaders(
+          authResponse,
+          'Failed to get cookie from SAML callback response'
+        );
+      }
+
+      // disable retry if not 5xx errors
+      if (authResponse.status < 500) {
+        attemptsLeft = -1;
+      }
+
+      throw new Error(`SAML callback failed: expected 302, got ${authResponse.status}`);
+    } catch (ex) {
+      cleanException(kbnHost, ex);
+      // Logging the `Cookie: sid=xxxx` header is safe here since it’s an intermediate, non-authenticated cookie that cannot be reused if leaked.
+      log.error(`Request sent: ${util.inspect(request)}`);
+
+      // exit for non 5xx errors
+      if (attemptsLeft === -1) {
+        throw ex;
+      }
+
+      // retry for 5xx errors
+      if (--attemptsLeft > 0) {
+        const attemptDelay = randomInt(500, 2_500);
```

Comment:
@azasypkin I was thinking to have a randomized delay so that in parallel threads (workers) we are not trying to hit API at the very same time. Does it make sense?